### PR TITLE
test: Add DB/Node Restart tests

### LINKED ISSUE #1082: No tests explicitly cover database restart
No tests explicitly cover database restart - the change detector does cover this implicitly, but only if a breaking change doc has not been provided.  I nearly missed a bug recently because of this, and we should add (non-change-detector) tests to cover this.

### PR DESCRIPTION
**Pull request by AndrewSisley**
## Relevant issue(s)

Resolves #1082 

## Description

Adds DB/Node Restart tests.  This will be required by Lens (e.g. to make sure migrations are not lost on restart). IIRC I think this also found https://github.com/sourcenetwork/defradb/pull/1482 (also fixed by Fred).

Fred was a massive help here and fixed the difficult bits when I flaked out on Friday :) Any commit that lists him as an author are his, I am only tagged in them having rebased them.

Commits are messy as anything as I wanted to preserve the level of credit due to Fred, and I'd suggest reviewing this PR at PR level as cleanup has not been squashed into commits introducing broken/messy code.

I will merge this when approved, whether that is 0.5.1 or 0.6 - it introduces no breaking (or risky) changes to the production code.

### CODE REVIEW

**Review by fredcarle — APPROVED** [review]
LGTM. Of course I'm biased here so I understand if you want to wait for someone else's approval.

**Comment by AndrewSisley** [comment]
> LGTM. Of course I'm biased here so I understand if you want to wait for someone else's approval.

:laughing: I thought I fixed the change detector earlier but hadn't - so the other have at least until then :grin:

**Comment by AndrewSisley** [comment]
> > LGTM. Of course I'm biased here so I understand if you want to wait for someone else's approval.
> 
> laughing I thought I fixed the change detector earlier but hadn't - so the other have at least until then grin

Was an existing change detector bug - this was just the first test to hit it.

**Pull request merged**
